File: python/grep/grep.py
def grep(pattern, flags, files):
    n, l, i, v, x = "-n" in flags, "-l" in flags, "-i" in flags, "-v" in flags, "-x" in flags
    if i: pattern = pattern.lower()
    multiple = len(files) > 1
    results = ""
    for f in files:
        for line_no, line in enumerate(open(f).readlines()):
            line_clean = line.lower() if i else line
            found = (x and pattern + "\n" == line_clean) or (not x and pattern in line_clean)
            if (found and not v) or (not found and v):
                if l:
                    results += f"{f}\n"
                    break
                results += (f"{f}:" if multiple else '') + (f"{line_no + 1}:" if n else '') + line
    return results

File: ruby/grep/grep.rb
class Grep
  def self.grep(pattern, flags, files)
    flags = flags.map{ |f| f[1].to_sym }.zip(flags.map { 1 }).to_h

    pattern = "^" + pattern + "$" if flags[:x]
    pattern = /#{pattern}/i       if flags[:i]

    out = []
    files.each { |f|
      File.read(f).split("\n").each_with_index { |line, i|
        if line.match?(pattern) ^ flags[:v]
          if flags[:l]
            out << f
            break
          end
          line = "#{(i + 1)}:" + line if flags[:n]
          line = "#{f}:" + line if files.size > 1
          out << line
        end
      }
    }
    out.join("\n")
  end
end
